### PR TITLE
Add Akamai EdgeGrid auth

### PR DESCRIPTION
**Pull request by vikiCoder**
- Added helper for Akamai's EdgeGrid authorization as requested in https://github.com/postmanlabs/postman-app-support/issues/6966

Todo:
- [x] Add auth helper for EdgeGrid
- [x] Add tests
- [x] Add changlog

### CODE REVIEW

File: lib/authorizer/edgegrid.js [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+     * Returns current timestamp in the format described in EdgeGrid specification (yyyyMMddTHH:mm:ss+0000)
+     *
+     * @returns {String} UTC timestamp in format yyyyMMddTHH:mm:ss+0000
+     */
+    getTimestamp = function () {
+        var timestamp,
+            date = new Date();
+
+        timestamp = date.getUTCFullYear() +
+                    String(date.getUTCMonth() + 1).padStart(2, ZERO) +
```
Comment (codenirvana) [inline lib/authorizer/edgegrid.js]: `padStart` requires Node v8+

File: lib/authorizer/edgegrid.js [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+                    String(date.getUTCDate()).padStart(2, ZERO) +
+                    DATE_TIME_SEPARATOR +
+                    String(date.getUTCHours()).padStart(2, ZERO) +
+                    COLON +
+                    String(date.getUTCMinutes()).padStart(2, ZERO) +
+                    COLON +
+                    String(date.getUTCSeconds()).padStart(2, ZERO) +
+                    UTC_OFFSET;
+
+        return timestamp;
```
Comment (codenirvana) [inline lib/authorizer/edgegrid.js]: can return directly if `timestamp` is not used elsewhere.

File: lib/authorizer/edgegrid.js [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+     * @param {String[]} headersToSign Headers to include in signature
+     * @param {Object} headers Request headers
+     * @returns {String} Canonicalized headers
+     */
+    canonicalizeHeaders = function (headersToSign, headers) {
+        var formattedHeaders = [],
+            headerValue;
+
+        headersToSign.forEach(function (headerName) {
+            headerName = headerName.trim().toLowerCase();
```
Comment (codenirvana) [inline lib/authorizer/edgegrid.js]: can `headerName` be null, undefined or non-string?

File: lib/authorizer/edgegrid.js [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+     * @param {String} params.timestamp Timestamp as defined in protocol specification
+     * @param {String} [params.bodyHash] Base64-encoded SHA–256 hash of request body for POST request
+     * @param {Object[]} params.headers Request headers
+     * @param {String[]} params.headersToSign Ordered list of headers to include in signature
+     * @param {String} params.method Request method
+     * @param {Url} params.url Node's URL object
+     * @returns {String} Authorization header
+     */
+    computeHeader: function (params) {
+        var authHeader = 'EG1-HMAC-SHA256 ',
```
Comment (codenirvana) [inline lib/authorizer/edgegrid.js]: top-level const?

File: lib/authorizer/edgegrid.js [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+            dataToSign;
+
+        authHeader += `client_token=${params.clientToken};`;
+        authHeader += `access_token=${params.accessToken};`;
+        authHeader += `timestamp=${params.timestamp};`;
+        authHeader += `nonce=${params.nonce};`;
+
+        dataToSign = [
+            params.method,
+            params.url.protocol.replace(COLON, EMPTY),
```
Comment (codenirvana) [inline lib/authorizer/edgegrid.js]: 3 level nested 😨 
make sure the necessary checks for `params`  was done before.

File: lib/authorizer/edgegrid.js [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+     * @returns {String} Canonicalized headers
+     */
+    canonicalizeHeaders = function (headersToSign, headers) {
+        var formattedHeaders = [],
+            headerValue;
+
+        headersToSign.forEach(function (headerName) {
+            if (typeof headerName !== 'string') { return; }
+
+            headerName = headerName.trim().toLowerCase();
```
Comment (codenirvana) [inline lib/authorizer/edgegrid.js]: why trim?

File: lib/authorizer/edgegrid.js [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+
+    /**
+     * Pads the given number with '0' and returns a string with given length (or greater if number is already longer
+     * than given length)
+     *
+     * @param {Number} num Number to pad
+     * @param {Number} length Minimum length of the string to return
+     * @returns {String} Padded number
+     */
+    padNumber = function (num, length) {
```
[the remark this answers is not in the document]
Comment (codenirvana) [reply lib/authorizer/edgegrid.js]: Also, why not lodash for this?

File: lib/authorizer/edgegrid.js [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+        params.baseURL = params.baseURL && urlEncoder.toNodeUrl(new sdk.Url(params.baseURL).toString(true)).host;
+        params.nonce = params.nonce || randomString(6);
+        params.timestamp = params.timestamp || getTimestamp();
+        params.url = url;
+        params.method = request.method;
+
+        // ensure that headers are case-insensitive as specified in the documentation
+        params.headers = request.getHeaders({enabled: true, ignoreCase: true});
+
+        if (typeof params.headersToSign === 'string') {
```
Comment (codenirvana) [inline lib/authorizer/edgegrid.js]: top-level const for `string`, used at multiple places.

File: lib/authorizer/edgegrid.js (hunk 6 above)
Comment (vikiCoder) [reply lib/authorizer/edgegrid.js]: To handle user input like ' X, Y, Z ' which will be like [' X', ' Y', ' Z '] after splitting by comma.

File: test/integration/auth-methods/edgegrid.test.js [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+        it('should have sent the request once', function () {
+            expect(testrun).to.nested.include({
+                'request.callCount': 1
+            });
+
+            var request = testrun.request.getCall(0).args[3],
+                response = testrun.request.getCall(0).args[2];
+
+            expect(request.url.toString()).to.eql(edgeGridAuthServer.url);
+            expect(response).to.have.property('code').that.is.oneOf([200, 401]);
```
Comment (codenirvana) [inline test/integration/auth-methods/edgegrid.test.js]: why `oneOf`?

File: test/integration/auth-methods/edgegrid.test.js [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+            });
+        });
+
+        it('should not include Authorization header', function () {
+            var request = testrun.request.getCall(0).args[3];
+
+            expect(request.headers.get('authorization')).to.be.undefined;
+        });
+
+        it('should have completed the run', function () {
```
Comment (codenirvana) [inline test/integration/auth-methods/edgegrid.test.js]: this should be the first assertion.

File: test/integration/auth-methods/edgegrid.test.js [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+
+        it('should include correct Authorization header', function () {
+            var request = testrun.request.getCall(0).args[3],
+                header = request.headers.members[0],
+                nonceRegex = /nonce=([A-Z]|[a-z]|[0-9]|-){36};/,
+                timestampRegex = /timestamp=[0-9]{8}T[0-9]{2}:[0-9]{2}:[0-9]{2}\+0000;/;
+
+            expect(header).to.have.have.property('key', 'Authorization');
+            expect(header).to.have.have.property('value').that.include(`client_token=${credentials.clientToken}`);
+            expect(header).to.have.have.property('value').that.include(`access_token=${credentials.accessToken}`);
```
Comment (codenirvana) [inline test/integration/auth-methods/edgegrid.test.js]: split these into multiple lines.
```suggestion
            expect(header).to.have.have.property('value')
              .that.include(`access_token=${credentials.accessToken}`);
```

File: test/unit/auth-handlers.test.js [hunk 12]
```diff
@@ -1211,4 +1211,294 @@ describe('Auth Handler:', function () {
...
+                headers = request.headers.all();
+
+                expect(headers).to.have.lengthOf(1);
+
+                authHeader = headers[0];
+
+                expect(authHeader.system).to.be.true;
+
+                authHeader = authHeader.toString();
+                expect(authHeader).to.include('client_token=');
```
Comment (codenirvana) [inline test/unit/auth-handlers.test.js]: can these be regex check with there expected value type

File: test/unit/auth-handlers.test.js [hunk 13]
```diff
@@ -1211,4 +1211,294 @@ describe('Auth Handler:', function () {
...
+            var rawReq = requestWithoutOptionalParams,
+                request = new Request(rawReq),
+                auth = request.auth,
+                authInterface = createAuthInterface(auth),
+                handler = AuthLoader.getHandler(auth.type),
+                headers,
+                authHeader;
+
+            handler.sign(authInterface, request, function () {
+                headers = request.headers.all();
```
Comment (codenirvana) [inline test/unit/auth-handlers.test.js]: headers and authHeader variables should be defined in this function scope.

**Pull request merged**